subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppIncrementalBuildIntegrationTest.groovy

```diff
@@ -0,0 +1,236 @@
+
+        def install = installation("app/build/install/main/${variant.toLowerCase()}")
+
+        when:
+        sourceFile.text = app.alternateMainSource.content
+
+        and:
+        run installApp
+
+        then:
+        skipped compileTasksDebug(LIBRARY)
+        skipped linkTaskDebug(LIBRARY)
+        executedAndNotSkipped compileTasksDebug(APP)
+        executedAndNotSkipped linkTaskDebug(APP)
+        executedAndNotSkipped installApp
+
+        and:
+        install.assertInstalled()
+        install.exec().out == app.alternateOutput
+    }
+
+    def "recompiles library and relinks executable with library source file change"() {
+        given:
+        run installApp
+        maybeWait()
+        def install = installation("app/build/install/main/debug")
+
+        when:
+        for (int i = 0; i < librarySourceFiles.size(); i++) {
+            TestFile sourceFile = librarySourceFiles.get(i)
+            sourceFile.text = app.alternateLibrarySources[i].content
+        }
+
+        and:
+        run installApp
+
+        then:
+        executedAndNotSkipped compileTasksDebug(LIBRARY)
+        executedAndNotSkipped linkTaskDebug(LIBRARY)
+        skipped compileTasksDebug(APP)
+        executedAndNotSkipped linkTaskDebug(APP)
+        executedAndNotSkipped installApp
+
+        and:
+        install.assertInstalled()
+        install.exec().out == app.alternateLibraryOutput
+    }
+
+    def "recompiles binary when header file changes"() {
+        given:
+        run installApp
+        maybeWait()
+
+        when:
+        headerFile << """
+            int unused();
+        """
+        run installApp
+
+        then:
+        executedAndNotSkipped compileTasksDebug(LIBRARY)
+        executedAndNotSkipped compileTasksDebug(APP)
+
+        if (nonDeterministicCompilation()) {
+            // Relinking may (or may not) be required after recompiling
+            executed linkTaskDebug(LIBRARY)
+            executed linkTaskDebug(APP), installApp
+        } else {
+            skipped linkTaskDebug(LIBRARY)
+            skipped linkTaskDebug(APP), installApp
+        }
+    }
+
+    def "recompiles binary when header file changes in a way that does not affect the object files"() {
+        given:
+        run installApp
+        maybeWait()
+
+        when:
+        headerFile << """
+            // Comment added to the end of the header file
+        """
+        run installApp
+
+        then:
+        executedAndNotSkipped compileTasksDebug(LIBRARY)
+        executedAndNotSkipped compileTasksDebug(APP)
+
+        if (nonDeterministicCompilation()) {
+            // Relinking may (or may not) be required after recompiling
+            executed linkTaskDebug(LIBRARY)
+            executed linkTaskDebug(APP), installApp
+        } else {
+            skipped linkTaskDebug(LIBRARY)
+            skipped linkTaskDebug(APP), installApp
+        }
+    }
+
+    def "recompiles binary when header file with relative path changes"() {
```

Comment:
These are incremental build tests for the new `cpp` plugins. The ones in `AbstractNativeLanguageIncrementalBuildIntegrationTest` are for the software model. It just didn't feel worth the effort to make the abstraction powerful enough to work with both plugins, so I ended up porting the tests to the new plugin. Should I invest some more time into abstracting this so the tests can be re-used?

Reply:
Since we are going to remove the software model stuff at some point, I think it's okay to duplicate the tests this way. It would make sense to have some abstraction at some point so that we don't have to repeat ourselves to support the new C plugin as well, but that doesn't have to happen right now.

Reply:
👍 Given that there is no new C plugin at this point we can postpone the extra abstraction.